### PR TITLE
Dataflow pipeline to match DLP SSN quotes to keyed hash dictionary

### PR DESCRIPTION
**Pull request by onetwopunch**
Adds a new Dataflow pipeline that does the following:

1. Listens on PubSub for filename changes in GCS
2. Chunks the file to submit to DLP API
3. For each SSN quote match to a Firestore of hashed, known SSNs
4. If there was an SSN match, emit the hash digest and filename where it was found to PubSub

Additional features:
* Terraform code to spin up Cloud Function, Bucket, PubSub and IAM.
* Python tool (hasher.py) to enable hashing and uploading hashed SSNs from the Datacenter to GCP
* Python tool (poller.py) to validate the results by counting the hashed findings. This also serves as an example of how to consume this Pubsub subscription.

### CODE REVIEW

**Comment by onetwopunch** [comment]
@prathapreddy123 I think you probably have the most context on this. Would you mind reviewing please?

**Comment by prathapreddy123** [comment]
@onetwopunch -  Lint checks are failing.  Can you please run local lint checks as specified [here](https://github.com/GoogleCloudPlatform/professional-services/blob/master/CONTRIBUTING.md) in Steps 3 and 4.

File: examples/dataflow-dlp-hash-pipeline/src/main/java/com/google/cloud/pso/hashpipeline/HashQuotesDoFn.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      LOG.error("Failed to create DLP Service Client", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c)
+      throws IOException, NoSuchAlgorithmException, InvalidKeyException {
+    KV<String, String> entry = c.element();
+    SecretKeySpec keySpec = new SecretKeySpec(this.hashkey, HMAC_SHA256);
```
Comment (prathapreddy123) [inline examples/dataflow-dlp-hash-pipeline/src/main/java/com/google/cloud/pso/hashpipeline/HashQuotesDoFn.java:66]: Initialize keyspec in setup and reuse the same across all elements

File: examples/dataflow-dlp-hash-pipeline/src/main/java/com/google/cloud/pso/hashpipeline/KVtoJSONDoFn.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import org.apache.beam.sdk.values.KV;
+
+public class KVtoJSONDoFn extends DoFn<KV<String, String>, String> {
+  @ProcessElement
+  public void processElement(ProcessContext c) {
+    KV<String, String> elem = c.element();
+    HashMap<String, String> map = new HashMap<String, String>();
+    map.put("filename", elem.getKey());
+    map.put("hashed_ssn", elem.getValue());
+    Gson gson = new Gson();
```
Comment (prathapreddy123) [inline examples/dataflow-dlp-hash-pipeline/src/main/java/com/google/cloud/pso/hashpipeline/KVtoJSONDoFn.java:33]: Create Gson instance in setup as same instance can be used across elements

File: examples/dataflow-dlp-hash-pipeline/src/test/java/com/google/cloud/pso/hashpipeline/ChunkFileDoFnTest.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@RunWith(JUnit4.class)
+public class ChunkFileDoFnTest {
+  private static final Logger log = LoggerFactory.getLogger(ChunkFileDoFnTest.class);
+  @Rule public final transient TestPipeline pipeline = TestPipeline.create();
+  @ClassRule public static TemporaryFolder tempFolder = new TemporaryFolder();
```
Comment (prathapreddy123) [inline examples/dataflow-dlp-hash-pipeline/src/test/java/com/google/cloud/pso/hashpipeline/ChunkFileDoFnTest.java:48]: Is this rule used?

File: examples/dataflow-dlp-hash-pipeline/infrastructure/hashpipeline-trigger/requirements.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+google-cloud-pubsub
```
Comment (prathapreddy123) [inline examples/dataflow-dlp-hash-pipeline/infrastructure/hashpipeline-trigger/requirements.txt:1]: Add version

File: examples/dataflow-dlp-hash-pipeline/infrastructure/hashpipeline-trigger/main.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# Copyright 2020 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
```
[the remark this answers is not in the document]
Comment (onetwopunch) [reply examples/dataflow-dlp-hash-pipeline/infrastructure/hashpipeline-trigger/main.py:3]: I believe we talked about this, but not quite sure. While you can have dataflow directly listen to GCS, there's two limitations:

1. Events only fire once the object is created, not updated
2. This forces you to create a new dataflow pipeline for each new bucket and templatize the bucket name. Possibly this is a common pattern though?

What are your thoughts?

File: examples/dataflow-dlp-hash-pipeline/README.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+* `terraform` executable is available in `$PATH`
+* `gcloud` is installed and up to date
+* `python` is version 3.5 or higher
+
+
+### Step 1: Deploy the Infrastructure
+First, ensure that the following APIs are enabled on your project:
+
+* `cloudfunctions.googleapis.com`
```
[the remark this answers is not in the document]
Comment (onetwopunch) [reply examples/dataflow-dlp-hash-pipeline/README.md:37]: Will just add to the terraform. I was just being lazy

**Comment by prathapreddy123** [comment]
Thanks @onetwopunch. LGTM

File: examples/dataflow-dlp-hash-pipeline/infrastructure/hashpipeline-trigger/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-dlp-hash-pipeline/infrastructure/hashpipeline-trigger/main.py]: Based on offline discussion the solution is changed to directly listen to GCS Pub/Sub Notifications and cloud function is eliminated

**Pull request merged**